File: Data Structures & Algorithms/eating-bananas/submission-5.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int r = piles[0];
        int l = 1;
        for (int i = 0; i < piles.size(); i++) {
            if (piles[i] > r) r = piles[i];
        }
        int res = r;

        while (l <= r) {
            int m = l + ((r - l) / 2);
            int t = 0;
            for (int i = 0; i < piles.size(); i++) {
                t += ceil(static_cast<double>(piles[i]) / m);
            }
            if (t <= h) {
                res = min(res, m);
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-5.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        res = nums[0]

        while l <= r:
            m = (l + r) // 2
            if nums[m] >= nums[l]:
                res = min(res, nums[l])
                l = m + 1
            elif nums[m] > nums[r]:
                l = m + 1
            else:
                r = m - 1
            res = min(res, nums[m])
        
        return res